feat(eventProcessor): Adds EventProcessor and BatchEventProcessor

mariamjamal94 opened the pull request:
Summary
-------

- Introduces an EventProcessor interface.
- Introduces a BatchEventProcessor

Buffering events within a queue before dispatching is an optimization that should prevent SDK implementations from exhausting resources while increasing throughput. This implementation relies on a BlockingCollection to buffer events received from one-to-many producers. A single consumer thread continuously polls from this queue to build a batch before emitting the batched LogEvent.

Test plan
---------

- Added unit tests.

Commit:
fix: remove unsupported package.

mariamjamal94 committed:
Update .travis.yml

mariamjamal94 committed:
Update .travis.yml

mariamjamal94 committed:
Update .travis.yml

mariamjamal94 committed:
Update .travis.yml

Commit:
Merge branch 'master' of https://github.com/Mariamjamal32/python-sdk

Commit:
Merge branch 'master' into mjamal/event_factory

Commit:
Merge branch 'mjamal/event_factory' of https://github.com/Mariamjamal32/python-sdk into mjamal/event_factory

# Conflicts:
#	.travis.yml

Commit:
fix: resolve issues in _create_visitor class method

Commit:
Merge branch 'master' of git://github.com/optimizely/python-sdk

Commit:
Merge branch 'master' into mjamal/event_factory

Commit:
feat: add event_processor interface to support batch event processing.

Commit:
Merge remote-tracking branch 'remotes/upstream/sohail/pr-193' into batch_event_processor

mnoman09 committed:
Merge branch 'sohail/pr-193' into batch_event_processor

Commit:
sync batch_event_processor with pr-193.

Commit:
Merge branch 'batch_event_processor' of https://github.com/Mariamjamal32/python-sdk into batch_event_processor

mnoman09 commented on optimizely/event/event_processor.py:
```diff
@@ -0,0 +1,207 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import threading
```
nit: Add space between comment and import

mnoman09 commented on optimizely/event/event_processor.py:
```diff
@@ -0,0 +1,207 @@
+from .event_factory import EventFactory
+
+
+class EventProcessor(object):
```
This class should be in separate file as this will be used in forward event processor as well and when we will add ForwardEventProcessor in this file, Then this file will get to long. 

By the way, I don't know what is the pattern we are using in python for adding separate classes but you can discuss with Sohail bhai.

msohailhussain replied:
I think it's not needed.

msohailhussain replied:
Can we extend from ABC. 
https://github.com/optimizely/python-sdk/blob/master/optimizely/config_manager.py#L29 but let me confirm first, i don't know the reason.

mnoman09 commented on optimizely/event/event_processor.py:
```diff
@@ -0,0 +1,207 @@
+  def __init__(self,
+                event_dispatcher,
+                logger,
+                start,
```
This should be set to False by default. because anyone can pass anything in it.

mnoman09 commented on optimizely/event/event_processor.py:
```diff
@@ -0,0 +1,207 @@
+    self._is_started = False
+    self._current_batch = list()
+
+    if start:
```
add isTrue check on it

mnoman09 commented on optimizely/event/event_processor.py:
```diff
@@ -0,0 +1,207 @@
+  def disposed(self):
+    return self._disposed
+
+  def _get_time_in_ms(self, _time=None):
```
I think we should put this function in some utility class, if python has any or we can make DateTimeUtil class so that in future when ever we need to use get_time_in_ms we can use it directly

mariamjamal94 replied:
For this function only?

mnoman09 reviewed (changes requested):
Looks good, just added few comments.

mariamjamal94 committed:
Update event_processor.py

Commit:
update: cater review comments.

msohailhussain commented on optimizely/event/event_processor.py:
```diff
@@ -0,0 +1,208 @@
+from datetime import timedelta
+from six.moves import queue
+
+from .entity.user_event import UserEvent
```
Please alphabetize, if it's the same pattern in python.

mariamjamal94 replied:
In python, we declare third party imports first alphabetically and then local imports alphabetically.

msohailhussain commented on optimizely/event/event_processor.py:
```diff
@@ -0,0 +1,208 @@
+  """ Class encapsulating event_processor functionality. Override with your own processor
+  providing process method. """
+
+  @staticmethod
```
why it's static, it shouldn't be static.

mariamjamal94 replied:
This is the convention being used in Python SDK to provide interface like functionality because python does not have interface structures of its own. The same convention can be found in error_handler.py.
We could use ABC module here and declare it as an abstract method.

msohailhussain commented on optimizely/event/event_processor.py:
```diff
@@ -0,0 +1,208 @@
+  def __init__(self,
+                event_dispatcher,
+                logger,
+                start=False,
```
can you just say defaultStart?

msohailhussain commented on optimizely/event/event_processor.py:
```diff
@@ -0,0 +1,208 @@
+        try:
+          item = self.event_queue.get(True, 0.05)
+
+        except queue.Empty:
```
Need to check performance, exception is not good at all. Why it can raise exception.

mariamjamal94 replied:
I could not find another way about it in Python.
Have a reference here: https://stackoverflow.com/questions/11247439/python-queue-empty-exception-handling
Please let me know what you think about it.

Commit:
update: review suggestions addressed.

Commit:
fix: address review comment to add abstract method from ABC in event_processor.

Commit:
update event/event_processor.py

Commit:
fix: remove stop and add close method in BatchEventProcessor.

Commit:
fix: reset indentation.

Commit:
update BatchEventProcessor.

Commit:
fix: remove stop from test_event_processor.

Commit:
fix: logger level.

Commit:
update: fallback to defaults for invalid and negative values BatchEventProcessor.

Commit:
fix: linter error.

mnoman09 commented on optimizely/event/event_processor.py:
```diff
@@ -63,9 +63,11 @@ def __init__(self,
-    self.batch_size = batch_size or self._DEFAULT_BATCH_SIZE
-    self.flush_interval = flush_interval or self._DEFAULT_FLUSH_INTERVAL
-    self.timeout_interval = timeout_interval or self._DEFAULT_TIMEOUT_INTERVAL
+    self.batch_size = batch_size if batch_size is not None and batch_size > 0 else self._DEFAULT_BATCH_SIZE
```
make sure that batch size is integer and also add its unit test

mnoman09 commented on optimizely/event/event_processor.py:
```diff
@@ -63,9 +63,11 @@ def __init__(self,
-    self.flush_interval = flush_interval or self._DEFAULT_FLUSH_INTERVAL
-    self.timeout_interval = timeout_interval or self._DEFAULT_TIMEOUT_INTERVAL
+    self.batch_size = batch_size if batch_size is not None and batch_size > 0 else self._DEFAULT_BATCH_SIZE
+    self.flush_interval = timedelta(milliseconds=flush_interval) if flush_interval is not None and flush_interval > 0 \
```
Same as above. that flush interval is number and add unit test

mnoman09 commented on optimizely/event/event_processor.py:
```diff
@@ -63,9 +63,11 @@ def __init__(self,
+    self.batch_size = batch_size if batch_size is not None and batch_size > 0 else self._DEFAULT_BATCH_SIZE
+    self.flush_interval = timedelta(milliseconds=flush_interval) if flush_interval is not None and flush_interval > 0 \
+                            else self._DEFAULT_FLUSH_INTERVAL
+    self.timeout_interval = timedelta(milliseconds=timeout_interval) if timeout_interval is not None and timeout_interval > 0 \
```
same as above.

mnoman09 commented on optimizely/event/event_processor.py:
```diff
@@ -0,0 +1,211 @@
+    try:
+      while True:
+        if self._get_time_in_ms() > self.flushing_interval_deadline:
+          self.logger.debug('Deadline exceeded; flushing current batch.')
```
we can remove this log as said by mike in ruby sdk PR [comment](https://github.com/optimizely/ruby-sdk/pull/191/files/096ae95ca7d8f68cf913b5c9583dccf22063db9d#r314019298)

mnoman09 reviewed (changes requested):
looks good just added few comments

mnoman09 commented on tests/test_event_processor.py:
```diff
@@ -0,0 +1,290 @@
+
+    self.assertStrictTrue(event_dispatcher.compare_events())
+    self._event_processor.close()
+
```
check this "self.assertEqual(0, self._event_processor.event_queue.qsize())" here as well

Commit:
update: review suggestion catered.

Commit:
update: import convention.

Commit:
update: remove uncertain log tests.

Commit:
update: handle 0ms flush_interval condition.

mnoman09 committed:
feat(eventProcessor): Adds EventProcessor and BatchEventProcessor

Summary
-------

- Introduces an EventProcessor interface.
- Introduces a BatchEventProcessor

Buffering events within a queue before dispatching is an optimization that should prevent SDK implementations from exhausting resources while increasing throughput. This implementation relies on a BlockingCollection to buffer events received from one-to-many producers. A single consumer thread continuously polls from this queue to build a batch before emitting the batched LogEvent.

Test plan
---------

- Added unit tests.

Commit:
update: _get_time_in_ms.

Commit:
Merge branch 'mnoman/AddBatchEP' of https://github.com/optimizely/python-sdk into batch_event_processor

Commit:
fix: flush_interval and batch_size default value.

Commit:
update: add default value logs.
